packages/utils/telemetry-utils/src/logger.ts

```diff
@@ -48,6 +50,33 @@ export interface ITelemetryLoggerPropertyBags{
     error?: ITelemetryLoggerPropertyBag;
 }
 
+// Less restrictive telemetry properties to include arrays with primitives.
+export interface ITelemetryPropertiesExt {
+    [index: string]: TelemetryEventPropertyType | ITaggedTelemetryPropertyType | TelemetryEventPropertyTypeExt;
+}
+
+// Base interface for logging telemetry statements allowing flat arrays containing primitives.
+export interface ITelemetryBaseEventExt extends ITelemetryPropertiesExt {
+    category: string;
+    eventName: string;
+}
+
+export interface ITelemetryGenericEventExt extends ITelemetryPropertiesExt {
+    eventName: string;
+    category?: TelemetryEventCategory;
+}
+export interface ITelemetryPerformanceEventExt extends ITelemetryGenericEventExt {
+    duration?: number; // Duration of event (optional)
+}
+
+export type TelemetryEventTypes =
```

Comment:
I see 3 solutions here:
1. We change core types and push responsibility to serialize arrays to consumers (hosts). This is breaking change as it impacts ITelemetryBaseLogger() and host implementation.
   - pros: much simpler types / handling, native support of arrays.
   - con: needs to be staged properly, pushes some complexity to hosts.
2. What you are doing
3. Same as above, but define unique types, not use unions. I.e. sendTelemetryEvent(event: ITelemetryGenericEventExt)

I prefer # 3 over # 2 - it makes it much easier to read code / follow types.
I'd say that it's worth going further - all of our code should use these types! So, either you follow up to replace all the usage in repo (where ITelemetryLogger interface is used, which is pretty much 99% of our code base) to these new types, or we reuse existing types to extend them, and create "legacy" (or rather - upstream types) for ITelemetryBaseLogger to be unchanged (has some implication on hosts).
Basically, we quite often propagate telemetry types through stack, and it would be pain for engineers to incrementally make these changes - I'd rather see this change done in one go.

I prefer # 1 over # 3. :) While it will take time to stage, I think change and final state will be much better.

Reply:
Please talk to Bohemia team RE this, as we would need some staging.
If there is pushback, what we can do also (not ideal, but workable solution) - is modified # 3, i.e. we create new set of interfaces that are to be used by ITelemetryBaseLogger and are copies of what we had before (such that host, for most part, does not need to do anything, other than using new names of interfaces), and reuse existing names to include arrays. TelemetryLogger can always json arrays to bridge the gap.

Please note that if you go on # 1, you need to be aware that staging (waiting) will be long. That's because version N of runtime may start to use arrays, and it maybe the version that app uses, but previous version of an app (like Chapter 1 app) does not have handling for arrays, but loads dynamically "future" version of container runtime and driver (that are on version N), which produce telemetry with arrays. That may break things.
So first, we need to ensure that all consumers can handle arrays. Wait some substantial amount of time for these changes to get to 100% prod saturation, and only then allow us to leverage arrays.
There are multiple ways to approach this problem. The one I'd suggest is the following:
- all the types are changed to support arrays, and we make a note in breaking.md that all apps have to deal with array
- but for some long period of time (multiple internal major releases), TelemetryLogger continuous to JSON arrays.
- and only after some reasonable amount of time, we remove that code and let hosting app layer to deal with arrays.